src/Parser.java

```diff
@@ -0,0 +1,255 @@
+import javafx.util.Pair;
+
+import java.util.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+public class Parser {
+    //lr(0)
+    private Grammar grammar;
+    private List<Set<Item>> states;
+    //pos state in states --> action name
+    //for reduce the action name is of the form:
+    //reduce,position in all productions from grammar,position in rules of production
+    private Map<Integer,String> action;
+    //pos state in states --> goto
+    private Map<Integer,Set<Goto>> goTo;
+
+    public Parser(Grammar grammar) throws Exception {
+        this.grammar = grammar;
+        action = new HashMap<>();
+        goTo = new HashMap<>();
+        collectionCanonical();
+        formTable();
+        System.out.println("The states are: ");
+         for(int i=0;i<states.size();i++){
+            System.out.println(i+"--"+states.get(i));
+        }
+        System.out.println("action=========");
+        System.out.println(action);
+        System.out.println("goto==========");
+        goTo.forEach((k,v)-> System.out.println(k+"---"+v));
+    }
+
+    //what a state contains
+    private Set<Item> closure(Set<Item> state){
+        /* Allocate space for the result. */
+        List<Item> setOfStates = new ArrayList<>(state);
```

Comment:
the logic here is ugly this way, maybe change it into a functional programming style with the help of the `Optional` class, and broken into multiple smaller methods with better naming scheme